packages/shopping-cart/src/shopping-cart-reducer.ts

```diff
@@ -85,29 +52,51 @@ function shouldQueueReducerEvent( cacheStatus: CacheStatus, action: ShoppingCart
 	return false;
 }
 
-function shoppingCartReducer(
+export function reducerWithQueue(
 	state: ShoppingCartState,
 	action: ShoppingCartAction
 ): ShoppingCartState {
-	const couponStatus = state.couponStatus;
+	if (
+		cacheStatusesForIgnoringReload.includes( state.cacheStatus ) &&
+		action.type === 'CART_RELOAD'
+	) {
+		debug( 'cart is pending an operation; ignoring reload action' );
+		return state;
+	}
 
 	// If the cacheStatus is 'fresh' or 'pending', then the initial cart has not
 	// yet loaded and so we cannot make changes to it yet. We therefore will
 	// queue any action that comes through during that time except for
 	// 'RECEIVE_INITIAL_RESPONSE_CART' or 'RAISE_ERROR'.
 	if ( shouldQueueReducerEvent( state.cacheStatus, action ) ) {
-		if ( action.type === 'CART_RELOAD' ) {
-			debug( 'cart has not yet loaded; ignoring reload action', action );
-			return state;
-		}
 		debug( 'cart has not yet loaded; queuing requested action', action );
 		return {
 			...state,
 			queuedActions: [ ...state.queuedActions, action ],
 		};
 	}
 
+	state = shoppingCartReducer( state, action );
+
+	if ( shouldPlayQueuedActions( state ) ) {
+		debug( 'playing queued actions', state.queuedActions );
+		const actions: ShoppingCartAction[] = [
+			...state.queuedActions,
+			{ type: 'CLEAR_QUEUED_ACTIONS' },
+		];
+		state = actions.reduce( shoppingCartReducer, state );
+		debug( 'queued actions are dispatched and queue is cleared' );
+	}
+
```

Comment:
The `lastValidResponseCart` state could be integrated into the main `state` and handled by a reducer here after playing queued actions:
```js
state = shoppingCartReducer( state, { type: 'UPDATE_LAST_VALID_CART' } );
```
and the reducer would have a `case` for a new action:
```js
case 'UPDATE_LAST_VALID_CART':
  if ( ! isStatePendingUpdate( state ) ) {
    return {
      ...state,
      lastValidResponseCart: convertTempResponseCartToResponseCart( state.responseCart ),
    };
  }
  return state;
```
The `lastValidResponseCart` is conceptually part of the state, there are no side-effects associated with it, so it doesn't need to be managed separately.